rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,223 @@
+    can be easily defined by a subgraph developer ultimately providing 
+    the tools necesssary for front end developers to create effective 
+    user interfaces without the need to middleware.
+     
+## Urgency
+
+Fulltext search filters are not essential to the value proposition of 
+the Graph Node, so the implementation is not urgent. However, several 
+high profile users have communicated that it may be a conversion blocker, 
+so it would be a valuable addition and should be added immediately. 
+
+## Terminology
+
+- `lexeme`: a basic lexical unit of a language, consisting of one word or 
+  several words, considered as an abstract unit, and applied to a family 
+  of words related by form or meaning.
+- `morphology (linguistics)`: the study of words, how they are formed, 
+  and their relationship to other words in the same language. 
+- `fulltext search index`: the result of lexical and morphological 
+  analysis (stemming) of a set of text documents.  It provides frequency 
+  and location for the language-specific stems found in the text documents 
+  being indexed. 
+
+## Detailed Design
+
+### Subgraph Definition
+
+Part of the power of fulltext search indexes are their flexibility, so 
+it is important to expose a simple interface to facilitate useful 
+applications of the index and ulimately aim to reduce the need to create 
+new subgraphs for the express purpose of updating fulltext search fields 
+for end-use cases. 
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+
+The proposed process of adding a fulltext search index to an Entity 
+involves adding a special field to that Entity in the GraphQL schema of 
+the subgraph's definition. The subgraph developer will create a new virtual 
+field of type String that is marked with the special `@FullTextSearch` directive. 
+The directive will have two required parameters for specifying the language
+ and the fields to be incorporated. Verification of the index definition 
+ will ensure that all fields referenced are valid String type fields on 
+ that entity. To be clear, the proposed interface will add the fulltext 
+ search definitions to the subgraph deployment hash, so adding or updating 
+ a fulltext search index will require an update to the subgraph.  
+ With subgraph composition it will be possible to easily create new subgraphs 
+ that add specific fulltext search indexes to an existing subgraph. 
+
+Example fulltext search field definition:
+```graphql
+type ExampleItem @entity {
+    id: ID!,
+    name: String!,
+    description: String!,
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @fullTextSearchIndex(language: "english", fields: [{name: "name", weight: 5}, {name: "description", weight: 3}, {name: "specs", weight: 1}])
```

Comment:
Some input on the directive name is welcome here! 
With the possibility in mind of there being additional custom indexes later,  an option I've considered is using a more general directive such as: `@index(type: String, param1: [String], param2: [String])`

Reply:
I much prefer the @index approach

Reply:
I'm not a big fan of the directive approach and even less so when on top of a virtual attribute whose only purpose is to support the functioning of an index.

Indexes and data models are separate concerns, and forcing storage level concerns into the application data model hurts re-usability and limits flexibility for the Indexer.

I think I'd rather see this go somewhere in the manifest under i.e., `suggested_indexes`, where the subgraph developer could also provide hints to other indexes that they think may be useful initially and Indexers can choose to incorporate into their decisions until they have enough data on query patterns to make that decision for themselves.

Reply:
Given the need to apply full-text search indexes that span multiple entities, I would like to update the proposal to have the index defined as a separate type in the schema or an extension to the query type. This way the return type can be a `union` or `interface` type that includes the affected entities and a new query will be generated rather than simply adding a new filter to the existing entity query.  
Including the index definition in the manifest rather then the schema would work too, though I personally prefer the schema. See my reasoning in response to your other comment below.